### PR TITLE
Improved handling of native quotes

### LINKED ISSUE #12073: [twitter] Image attachments in shared tweets are shown twice
See 2022-10-29 14 36 11 friendica mrpetovan com e915340e3686

What can I check? "Item Source" in the admin panel doesn't show anything.

### PR DESCRIPTION
**Pull request by annando**
We are now only storing non native quotes in the body. All native quotes are done via the quote-uri-id field in the post. The whole code is cleaned up around this.

This PR has to be merged at the same time as the addon PR.

### CODE REVIEW

File: src/Content/Item.php [hunk 1]
```diff
@@ -796,4 +733,24 @@ private function getShareArray(array $item): array
...
+	 */
+	public function addShareLink(string $body, int $quote_uri_id): string
+	{
+		$post = Post::selectFirstPost(['uri', 'plink'], ['uri-id' => $quote_uri_id]);
+		if (empty($post)) {
+			return $body;
+		}
+
+		$link = $post['plink'] ?: $post['uri'];
+		$body .= "\n♲ " . $link;
```
Comment (MrPetovan) [inline src/Content/Item.php]: Energy is expensive, let's save the creation of an intermediary variable.
```suggestion
		$body .= "\n♲ " . ($post['plink'] ?: $post['uri']);
```

File: src/Model/Item.php [hunk 2]
```diff
@@ -3630,12 +3646,56 @@ protected static function isAllowedByUser(array $item, int $user_id): bool
...
+			Logger::debug('Fetched shared post', ['id' => $id, 'url' => $url, 'uid' => $uid]);
+
+			$shared_item = Post::selectFirst(['uri-id'], ['id' => $id]);
+			if (!DBA::isResult($shared_item)) {
+				Logger::warning('Post does not exist.', ['id' => $id, 'url' => $url, 'uid' => $uid]);
+				return 0;
+			}
+		}
+
+		return $shared_item['uri-id'];
```
Comment (MrPetovan) [inline src/Model/Item.php]: I'd rather we used early returns than the repetitive `empty($shared_item['uri-id'])` condition which means something different every time.
```suggestion
		if (!empty($shared['guid'])) {
			$shared_item = Post::selectFirst(['uri-id'], ['guid' => $shared['guid'], 'uid' => [0, $uid]]);
			if (!empty($shared_item['uri-id'])) {
				Logger::debug('Found post by guid', ['guid' => $shared['guid'], 'uid' => $uid]);
				return $shared_item['uri-id'];
			}
		}

		if (!empty($shared['message_id'])) {
			$shared_item = Post::selectFirst(['uri-id'], ['uri' => $shared['message_id'], 'uid' => [0, $uid]]);
			if (!empty($shared_item['uri-id'])) {
				Logger::debug('Found post by message_id', ['message_id' => $shared['message_id'], 'uid' => $uid]);
				return $shared_item['uri-id'];
			}
		}

		if (!empty($shared['link'])) {
			$shared_item = Post::selectFirst(['uri-id'], ['plink' => $shared['link'], 'uid' => [0, $uid]]);
			if (!empty($shared_item['uri-id'])) {
				Logger::debug('Found post by link', ['link' => $shared['link'], 'uid' => $uid]);
				return $shared_item['uri-id'];
			}
		}

		$url = $shared['message_id'] ?: $shared['link'];
		$id = self::fetchByLink($url);
		if (!$id) {
			Logger::notice('Post could not be fetched.', ['url' => $url, 'uid' => $uid]);
			return 0;
		}

		Logger::debug('Fetched shared post', ['id' => $id, 'url' => $url, 'uid' => $uid]);

		$shared_item = Post::selectFirst(['uri-id'], ['id' => $id]);
		if (!DBA::isResult($shared_item)) {
			Logger::warning('Post does not exist.', ['id' => $id, 'url' => $url, 'uid' => $uid]);
			return 0;
		}

		return $shared_item['uri-id'];
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Comment by annando** [comment]
@MrPetovan changes are applied.

File: src/Model/Item.php [hunk 3]
```diff
@@ -3630,12 +3646,57 @@ protected static function isAllowedByUser(array $item, int $user_id): bool
...
+
+		$shared_item = Post::selectFirst(['uri-id'], ['id' => $id]);
+		if (!DBA::isResult($shared_item)) {
+			Logger::warning('Post does not exist.', ['id' => $id, 'url' => $url, 'uid' => $uid]);
+			return 0;
+		} else {
+			Logger::debug('Fetched shared post', ['id' => $id, 'url' => $url, 'uid' => $uid]);
+		}
+
+		return $shared_item['uri-id'];
```
Comment (MrPetovan) [inline src/Model/Item.php]: Almost there, no need for an `else` statement here and we can avoid a negative condition.
```suggestion
		if (DBA::isResult($shared_item)) {
			Logger::debug('Fetched shared post', ['id' => $id, 'url' => $url, 'uid' => $uid]);
			return $shared_item['uri-id'];
		}

		Logger::warning('Post does not exist.', ['id' => $id, 'url' => $url, 'uid' => $uid]);
		return 0;
```

**Review by MrPetovan — CHANGES_REQUESTED** [review]

**Comment by MrPetovan** [comment]
Does this fix #12073 ?

**Comment by annando** [comment]
Yeah

**Comment by MrPetovan** [comment]
Did you see my second Change Request?

**Review by MrPetovan — APPROVED** [review]

**Pull request merged**